### PR TITLE
Issue17 pwm out for motor speeds

### PR DESCRIPTION
**Pull request by rkeen9**
Initialized PWM for motor speed control and created a function that takes a motor, speed percentage, and direction as its arguments. The function then powers the specified motor the speed percentage and direction.

### CODE REVIEW

File: Motor_Controller_2.0 Run (1).cfg [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+# STlink Debug clock frequency
+set CLOCK_FREQ 4000
+
+# use hardware reset, connect under reset
+# connect_assert_srst needed if low power mode application running (WFI...)
+reset_config srst_only srst_nogate connect_assert_srst
+set CONNECT_UNDER_RESET 1
+
+source [find target/stm32f4x.cfg]
```
Comment (lymacasm) [inline Motor_Controller_2.0 Run (1).cfg:28]: This file is generated by the build and shouldn't be under version control. It should be removed.

File: src/main.c [hunk 2]
```diff
@@ -85,7 +86,15 @@ int main(void)
...
-
+	init_motors();
+	motor_set_speed_percent(Motor1, 50, Forward); //PA1
+	motor_set_speed_percent(Motor2, 50, Forward); //PA3
+	motor_set_speed_percent(Motor3, 50, Forward); //PA10
+	motor_set_speed_percent(Motor4, 50, Forward); //PA8
+	motor_set_speed_percent(Motor5, 50, Forward); //PD14
+	motor_set_speed_percent(Motor6, 50, Forward); //PD12
+	motor_set_speed_percent(Motor7, 50, Forward); //PA6
+	motor_set_speed_percent(Motor8, 50, Forward); //PB0
```
Comment (lymacasm) [inline src/main.c:97]: Why are the motors set to 50?

File: src/motors.c [hunk 3]
```diff
@@ -56,6 +150,33 @@ extern void init_motors()
...
+			TIM_SetCompare4(TIM4, speed*10);
+			break;
+		case Motor6:
+			TIM_SetCompare1(TIM4, speed*10);
+			break;
+		case Motor7:
+			TIM_SetCompare1(TIM3, speed*10);
+			break;
+		case Motor8:
+			TIM_SetCompare3(TIM3, speed*10);
```
Comment (lymacasm) [inline src/motors.c:177]: uint8_t max number is 2^8=255, the max input speed we'd receive is 100, 100*10=1000 > max number for uint8_t. Not entirely sure what the input parameters to that function is, but if there may be an overflow we'd have to watch out for.

**Review by lymacasm — CHANGES_REQUESTED** [review]
Looks good! Extra file needs to be removed, and I have some questions/which may require changes.

File: src/motors.c (hunk 3 above)
Comment (batesandy123) [reply src/motors.c:177]: The max compare value is 1000. It's the max value of the capture compare register.

Still not sure why it is timed by 10? If you ask me this should take in raw speed, we can have another function or use the FSM to convert 0-255 to what it is in 1000.

File: src/motors.c [hunk 4]
```diff
@@ -29,7 +29,101 @@ static void init_motor_current_temp_MUX(){
...
+	GPIO_PinAFConfig(GPIOA, GPIO_PinSource3, GPIO_AF_TIM2);
+	GPIO_PinAFConfig(GPIOA, GPIO_PinSource10, GPIO_AF_TIM1);
+	GPIO_PinAFConfig(GPIOA, GPIO_PinSource8, GPIO_AF_TIM1);
+	GPIO_PinAFConfig(GPIOD, GPIO_PinSource15, GPIO_AF_TIM4);
+	GPIO_PinAFConfig(GPIOD, GPIO_PinSource12, GPIO_AF_TIM4);
+	GPIO_PinAFConfig(GPIOA, GPIO_PinSource6, GPIO_AF_TIM3);
+	GPIO_PinAFConfig(GPIOB, GPIO_PinSource0, GPIO_AF_TIM3);
+	//timer set up
+	TIM_TimeBaseInitTypeDef TIM_TimeBaseStructure;
+	TIM_TimeBaseStructure.TIM_Period = 1000 - 1;
```
Comment (batesandy123) [inline src/motors.c:76]: More resolution would be better. In reason we don't make this 10000 or 50000?

File: src/motors.c [hunk 5]
```diff
@@ -56,6 +150,33 @@ extern void init_motors()
 /* Robert's extern/visible functions go below here: */
 extern void motor_set_speed_percent(motors_t motor_x, uint8_t speed, direction_t dir)
 {
+	if(dir == Reverse) speed *= -1;
```
Comment (batesandy123) [inline src/motors.c:153]: So 9% duty cycle is neutral, 6% is full reverse and 12% is full forward. Thus we need to make reverse and forward reflect this.

File: src/motors.c (hunk 4 above)
Comment (rkeen9) [reply src/motors.c:76]: How would I modify the period without affecting the value that will be inputted into the capture compare register?

Comment (batesandy123) [reply src/motors.c:76]: You would have to scale them proportionally. You could skip it for now. We could make a new issue and comeback to it

**Pull request merged**